dspace-api/src/main/java/org/dspace/app/profile/ResearcherProfileServiceImpl.java

```diff
@@ -0,0 +1,338 @@
+import org.dspace.discovery.indexobject.IndexableCollection;
+import org.dspace.eperson.EPerson;
+import org.dspace.eperson.Group;
+import org.dspace.eperson.service.GroupService;
+import org.dspace.services.ConfigurationService;
+import org.dspace.util.UUIDUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.util.Assert;
+
+/**
+ * Implementation of {@link ResearcherProfileService}.
+ *
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
+ *
+ */
+public class ResearcherProfileServiceImpl implements ResearcherProfileService {
+
+    private static Logger log = LoggerFactory.getLogger(ResearcherProfileServiceImpl.class);
+
+    @Autowired
+    private ItemService itemService;
+
+    @Autowired
+    private WorkspaceItemService workspaceItemService;
+
+    @Autowired
+    private InstallItemService installItemService;
+
+    @Autowired
+    private ConfigurationService configurationService;
+
+    @Autowired
+    private CollectionService collectionService;
+
+    @Autowired
+    private SearchService searchService;
+
+    @Autowired
+    private GroupService groupService;
+
+    @Autowired
+    private AuthorizeService authorizeService;
+
+    @Autowired(required = false)
+    private List<AfterResearcherProfileCreationAction> afterCreationActions;
+
+    @PostConstruct
+    public void postConstruct() {
+
+        if (afterCreationActions == null) {
+            afterCreationActions = Collections.emptyList();
+        }
+
+    }
+
+    @Override
+    public ResearcherProfile findById(Context context, UUID id) throws SQLException, AuthorizeException {
+        Assert.notNull(id, "An id must be provided to find a researcher profile");
+
+        Item profileItem = findResearcherProfileItemById(context, id);
+        if (profileItem == null) {
+            return null;
+        }
+
+        return new ResearcherProfile(profileItem);
+    }
+
+    @Override
+    public ResearcherProfile createAndReturn(Context context, EPerson ePerson)
+            throws AuthorizeException, SQLException, SearchServiceException {
+
+        Item profileItem = findResearcherProfileItemById(context, ePerson.getID());
+        if (profileItem != null) {
+            throw new ResourceAlreadyExistsException("A profile is already linked to the provided User");
+        }
+
+        Collection collection = findProfileCollection(context)
+            .orElseThrow(() -> new IllegalStateException("No collection found for researcher profiles"));
+
+        context.turnOffAuthorisationSystem();
+        Item item = createProfileItem(context, ePerson, collection);
+        context.restoreAuthSystemState();
+
+        ResearcherProfile researcherProfile = new ResearcherProfile(item);
+
+        for (AfterResearcherProfileCreationAction afterCreationAction : afterCreationActions) {
+            afterCreationAction.perform(context, researcherProfile, ePerson);
```

Comment:
@LucaGiamminonni : The reason I asked for this is that this is an "After" creation callback sort of script.  The Researcher Profile has already been successfully created at this point.  However, if this `perform()` method throws an exception, it will cause everything to rollback (at least that's my reading of the code).  

All that said, if you feel it's appropriate in this case for the afterCreation action to have the ability to rollback all changes, that's fine with me.  But, we may want to have a test to prove that the Researcher Profile is not created if this perform throws an exception (that way we don't accidentally change this behavior in the future without thinking things through)